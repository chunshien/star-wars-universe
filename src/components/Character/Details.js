import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import styled from 'styled-components'

const Content = styled.div`
    background-color: white;
    box-sizing: border-box;
    padding: 20px;
    margin: 20px 20%;
    border-radius: 5px;
    box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14), 0px 1px 10px 0px rgba(0,0,0,0.12);
`;

const Field = styled.div`
    width: 100%;
    display: table;
`;

const Label = styled.div`
    width: 30%;
    display: table-cell;
    padding: 5px;
    text-align: right;
`;

const Value = styled.div`
    width: 70%;
    display: table-cell;
    padding: 5px;
`;


function Details(props) {
    const renderField = () => {
        const {details} = props;
        if (details) {
            return Object.keys(details).map((item)=>{
                const label = _.startCase(item.replace('_', ' '));
                if (typeof details[item] === 'object') {
                    return (
                        <Field key={item} >
                            <Label>{label} : </Label>
                            <Value>
                                {details[item].length > 0 ? details[item].map((child) => (
                                    <div key={child}>{child}</div>
                                )) : '-'}
                            </Value>
                        </Field>
                    )
                }
                else {
                    return (
                        <Field key={item}>
                            <Label>{label} : </Label>
                            <Value>{details[item]}</Value>
                        </Field>
                    )
                }
            });
        }
    }

    return (
        <Fragment>
            <Content>
                {renderField()}
            </Content>            
        </Fragment>
    )
}
Details.propTypes = {
    details: PropTypes.object
};
export default Details;