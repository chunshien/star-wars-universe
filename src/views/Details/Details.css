.details-header {
    text-align: center;
    font-size: 36px;
    font-weight: bold;
    padding: 10px;
}

.details-content {    
    background-color: white;
    box-sizing: border-box;
    padding: 20px;
    margin: 20px 20%;
    border-radius: 5px;
    box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14), 0px 1px 10px 0px rgba(0,0,0,0.12);
}

.details-field {
    width: 100%;
    display: table;
}

.details-label {
    width: 30%;
    display: table-cell;
    padding: 5px;
    text-align: right;
}

.details-value {
    width: 70%;
    display: table-cell;
    padding: 5px;
}

.back-button-container {
    padding: 10px 20%;    
}