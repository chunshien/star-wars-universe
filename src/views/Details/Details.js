import React, { Fragment, useEffect, useCallback } from 'react';
import _ from 'lodash';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import { getDetails } from '../../actions';
import { useHistory } from 'react-router-dom'

import './Details.css';
import Button from '../../components/Button/Button';

function Details({loading, getDetails, details}) {
    const { type, id } = useParams();
    let history = useHistory()  

    useEffect(() => {
        getDetails(type + '/' + id);
    }, [type, id, getDetails]);
    
    const renderField = () => {
        if (details) {
            return Object.keys(details).map((item)=>{
                const label = _.startCase(item.replace('_', ' '));
                if (typeof details[item] === 'object') {
                    return (
                        <div key={item} className={'details-field'}>
                            <div className={'details-label'}>{label} : </div>
                            <div className={'details-value'}>
                                {details[item].length > 0 ? details[item].map((child) => (
                                    <div key={child}>{child}</div>
                                )) : '-'}
                            </div>
                        </div>
                    )
                }
                else {
                    return (
                        <div key={item} className={'details-field'}>
                            <div className={'details-label'}>{label} : </div>
                            <div className={'details-value'}>{details[item]}</div>
                        </div>
                    )
                }
            });
        }
    }

    const handleBack = useCallback(() => {history.goBack()}, [history]);
    
    return (
        <Fragment>
            {loading && 
                <div className={'loader-container'}>
                    <div className={'loader'}></div>
                </div>
            }
            <div className={'details-header'}>Details</div>
            <div className={'details-content'}>
                {renderField()}
            </div>
            <div className={'back-button-container'}>
                <Button onClick={handleBack} />
            </div>
        </Fragment>
    )
}

const mapDispatchToProps = {
    getDetails: getDetails,
};

const mapStateToProps = (state) => ({
    details: state.details,
    loading: state.loading
})

export default  connect(mapStateToProps, mapDispatchToProps)(Details);