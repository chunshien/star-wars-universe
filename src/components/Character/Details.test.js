import React from 'react';
import renderer from 'react-test-renderer';

import Details from './Details';
const testData = require('../../assets/test-data/details')

describe('Component - Details Test', () => {
    it('renders Details without data without crashing', () => {
        const component = renderer.create(
            <Details/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders Details with data without crashing', () => {
        const component = renderer.create(
            <Details
                details={testData}                           
            />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
})