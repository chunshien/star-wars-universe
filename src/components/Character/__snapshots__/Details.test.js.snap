// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Details Test renders Details with data without crashing 1`] = `
<div
  className="sc-AxjAm jfFdTm"
>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Name
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      Luke Skywalker
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Height
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      172
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Mass
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      77
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Hair Color
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      blond
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Skin Color
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      fair
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Eye Color
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      blue
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Birth Year
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      19BBY
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Gender
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      male
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Homeworld
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      https://swapi.co/api/planets/1/
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Films
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      <div>
        https://swapi.co/api/films/2/
      </div>
      <div>
        https://swapi.co/api/films/6/
      </div>
      <div>
        https://swapi.co/api/films/3/
      </div>
      <div>
        https://swapi.co/api/films/1/
      </div>
      <div>
        https://swapi.co/api/films/7/
      </div>
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Species
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      <div>
        https://swapi.co/api/species/1/
      </div>
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Vehicles
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      <div>
        https://swapi.co/api/vehicles/14/
      </div>
      <div>
        https://swapi.co/api/vehicles/30/
      </div>
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Starships
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      <div>
        https://swapi.co/api/starships/12/
      </div>
      <div>
        https://swapi.co/api/starships/22/
      </div>
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Created
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      2014-12-09T13:50:51.644000Z
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Edited
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      2014-12-20T21:17:56.891000Z
    </div>
  </div>
  <div
    className="sc-AxirZ hrootq"
  >
    <div
      className="sc-AxiKw cscMYO"
    >
      Url
       : 
    </div>
    <div
      className="sc-AxhCb gsDaNY"
    >
      https://swapi.co/api/people/1/
    </div>
  </div>
</div>
`;

exports[`Component - Details Test renders Details without data without crashing 1`] = `
<div
  className="sc-AxjAm jfFdTm"
/>
`;
